### PR TITLE
Performance function name in scatterplots

### LINKED ISSUE #31: Performance function name in scatterplots
1. Pass the raw Python name of the performance function used to calculate model performance from the back end to the UI via a prop.

2. Display the performance function name in the scatter plots based on the prop value.

### PR DESCRIPTION
**Pull request by Nking92**
Displays the performance function name in the scatterplot.

image

- The performance metric name is now a prop on the PerformanceCompatibility component
- Performance metric can be specified with CompatibilityAnalysis constructor
- Rewrote and moved around some code to pass around performance_metric and fill it in with a default accuracy function if it was not specified by the user

Closes #31

### CODE REVIEW

**Review by nushib — APPROVED** [review]

File: backwardcompatibilityml/sweep_management.py [hunk 1]
```diff
@@ -114,8 +118,12 @@ def get_sweep_status(self):
         return self.last_sweep_status
 
     def get_sweep_summary(self):
+        metric_name = self.performance_metric.__name__
+        if metric_name == "model_accuracy":
```
Comment (ilmarinen) [inline backwardcompatibilityml/sweep_management.py:122]: There's no need to do this if check, because we always want to see the actual function name here.

**Review by ilmarinen — CHANGES_REQUESTED** [review]
Left one comment.

File: widget/PerformanceCompatibility.tsx [hunk 2]
```diff
@@ -171,7 +171,7 @@ class PerformanceCompatibility extends Component<PerformanceCompatibilityProps,
       .append('text')
         .attr('id', 'yAxisLabel')
         .attr('transform','rotate(-90)')
-        .attr('x',-h/2)
+        .attr('x',-h/2+2.5*this.props.performanceMetric.length)
```
Comment (Nking92) [inline widget/PerformanceCompatibility.tsx:174]: Added this to center the y-axis label for any string.

File: backwardcompatibilityml/helpers/training.py [hunk 3]
```diff
@@ -646,7 +604,7 @@ def compatibility_sweep(sweeps_folder_path, number_of_epochs, h1, h2,
                         training_set, test_set, batch_size_train, batch_size_test,
                         OptimizerClass, optimizer_kwargs,
                         NewErrorLossClass, StrictImitationLossClass,
-                        performance_metric=None,
+                        performance_metric=model_accuracy,
```
Comment (ilmarinen) [inline backwardcompatibilityml/helpers/training.py:607]: Thank you for the cleanup.

File: backwardcompatibilityml/metrics.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+import torch
+from sklearn.metrics import accuracy_score
+
+
+def model_accuracy(model, dataset, device="cpu"):
```
Comment (ilmarinen) [inline backwardcompatibilityml/metrics.py:8]: The one tradeoff with this approach is that the user will have to wrap their performance function to allow its function signature to be the same as this, and pass the wrapped function in.

Could we document this somewhere?

Two places I can think of are:

1. In the docstring of the compatibility_sweeps function.
2. Another place would be on the widget so that it is documented what type of function the user passes in.

**Review by ilmarinen — CHANGES_REQUESTED** [review]
Added another comment. Basically a request for documentation.

Otherwise looks good. Will approve after this.

File: backwardcompatibilityml/metrics.py (hunk 4 above)
Comment (Nking92) [reply backwardcompatibilityml/metrics.py:8]: I added docs to the compatibility_sweep function and to the CompatibilityAnalysis constructor. Here's what a user will see if they run `?CompatibilityAnalysis` in their notebook. Do you think it's sufficient? I'm not completely sure where to put this information on the widget itself without adding a lot of clutter.

image

File: backwardcompatibilityml/widget/compatibility_analysis.py [hunk 5]
```diff
@@ -169,6 +170,13 @@ class CompatibilityAnalysis(object):
         StrictImitationLossClass: The class of the Strict Imitation style loss
             function to be instantiated and used to perform compatibility
             constrained training of our model h2.
+        performance_metric: A function to evaluate model performance. The function is
```
Comment (ilmarinen) [inline backwardcompatibilityml/widget/compatibility_analysis.py:173]: Thanks.

**Review by ilmarinen — APPROVED** [review]
@Nking92 +1

**Review by jlema — APPROVED** [review]
Would prefer if the default performance metric function is displayed as Accuracy (or Model Accuracy) instead of model_accuracy but approved otherwise.

**Comment by ilmarinen** [comment]
> Would prefer if the default performance metric function is displayed as Accuracy (or Model Accuracy) instead of model_accuracy but approved otherwise.

@jlema As mentioned in chat. Using the programmatic name was a request from @nushib.
